usr/src/uts/common/fs/zfs/zil.c

```diff
@@ -1208,22 +1208,17 @@ zil_lwb_write_issue(zilog_t *zilog, lwb_t *lwb)
 	tx = dmu_tx_create(zilog->zl_os);
 
 	/*
-	 * Since we are not going to create any new dirty data and we can even
-	 * help with clearing the existing dirty data, we should not be subject
-	 * to the dirty data based delays.
-	 * We (ab)use TXG_WAITED to bypass the delay mechanism.
-	 * One side effect from using TXG_WAITED is that dmu_tx_assign() can
-	 * fail if the pool is suspended.  Those are dramatic circumstances,
-	 * so we return NULL to signal that the normal ZIL processing is not
-	 * possible and txg_wait_synced() should be used to ensure that the data
-	 * is on disk.
+	 * Since we are not going to create any new dirty data and we
+	 * can even help with clearing the existing dirty data, we
+	 * should not be subject to the dirty data based delays. We
+	 * (ab)use TXG_WAITED to bypass the delay mechanism.
 	 */
-	error = dmu_tx_assign(tx, TXG_WAITED);
-	if (error != 0) {
-		ASSERT3S(error, ==, EIO);
-		dmu_tx_abort(tx);
-		return (NULL);
+	while ((error = dmu_tx_assign(tx, TXG_WAITED)) != 0) {
+		VERIFY3S(error, ==, ERESTART);
+		dmu_tx_unassign(tx);
```

Comment:
It seems that `dmu_tx_unassign` is currently internal to dmu_tx.c and the build complains about that.

Reply:
And it seems that if `dmu_tx_assign` returns an error, then `dmu_tx_unassign` has already been called by it.

Reply:
I'll fix the build issues; I hadn't tested prior to opening the PR.